### PR TITLE
Extent elimination refactoring and extent map upgrade from v4 to v5

### PR DESCRIPTION
**Pull request by tntnatbry**

### CODE REVIEW

**Comment by tntnatbry** [comment]
@dhall-MariaDB Can you review the extent map upgrade code from v4 to v5 in versioning/BRM/extentmap.h(.cpp) along with @drrtuy ?

File: dbcon/joblist/lbidlist.cpp
Comment (drrtuy) [inline dbcon/joblist/lbidlist.cpp:298]: It is safe but sUnsigned from CSC namespace returns false for DECIMAL.

**Review by davidhall586 — APPROVED** [review]

File: dbcon/joblist/lbidlist.cpp
Comment (drrtuy) [reply dbcon/joblist/lbidlist.cpp:298]: *isUnsigned

**Review by drrtuy — CHANGES_REQUESTED** [review]
Don't merge it yet.

File: dbcon/joblist/lbidlist.cpp
Comment (drrtuy) [inline dbcon/joblist/lbidlist.cpp:711]: Will we compare wide-DECIMALs getting into else branch?

Comment (drrtuy) [reply dbcon/joblist/lbidlist.cpp:711]: I think the first question is: if it is possible to get here with wide-DECIMALs in the first place.

File: dbcon/joblist/lbidlist.cpp
Comment (drrtuy) [inline dbcon/joblist/lbidlist.cpp]: This approach surely works. 
But as times goes by the vision had changed and I see this as a 16-bytes read from BS::buf() and subsequent int128 cast though. It's up to you to decide should we proceed with this version or not.

File: dbcon/joblist/lbidlist.cpp
Comment (drrtuy) [inline dbcon/joblist/lbidlist.cpp:972]: I believe the code doesn't compile but just out of curiosity did you try to compile it w/o this explicit template specialization?

Comment (drrtuy) [reply dbcon/joblist/lbidlist.cpp:972]: The reason why I'm asking is that there is only one version of GetMinMax and that is templated. linker might ask for an explicit template to link against it though.

File: dbcon/joblist/rowestimator.cpp
Comment (drrtuy) [inline dbcon/joblist/rowestimator.cpp:149]: It seems like an automated change has been applied here. 
We return uint32_t from this f() into uint32_t variable so ret should be uint32_t not even uint64_t. Given the calculations this should be safe.

File: dbcon/joblist/rowestimator.cpp
Comment (drrtuy) [inline dbcon/joblist/rowestimator.cpp]: Could you ellaborate on this cast?

File: dbcon/joblist/rowestimator.cpp
Comment (drrtuy) [inline dbcon/joblist/rowestimator.cpp]: I presume here math works w/o an explicit cast to float128?

File: dbcon/joblist/tuple-bps.cpp
Comment (drrtuy) [inline dbcon/joblist/tuple-bps.cpp]: I think we should replace this with a constant from datatypes::

Comment (drrtuy) [reply dbcon/joblist/tuple-bps.cpp]: We should reduce a number of hardcoded values around the codebase I believe.

File: dbcon/joblist/tuple-bps.cpp
Comment (drrtuy) [inline dbcon/joblist/tuple-bps.cpp:1706]: Who is the author of this. I wish him to work with such condition expressions all his life.

File: dbcon/mysql/ha_mcs_partition.cpp
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_partition.cpp]: Could you replace this hardcoded value with a corresponding const from datatypes:: namespace from decimal.h or I can do this after the merge.

File: dbcon/joblist/lbidlist.cpp
Comment (tntnatbry) [reply dbcon/joblist/lbidlist.cpp:298]: Keeping this here for now, in case we add UDECIMAL to isUnsigned in the future.

File: dbcon/joblist/lbidlist.cpp
Comment (tntnatbry) [reply dbcon/joblist/lbidlist.cpp:711]: Yes. The calling function, TupleBPS::addCPPredicates(), filters out extents of the large-side table, based on the CP info of the small-side tables. So if either the small-side or the large-side table contains wide-decimal columns as join keys (which we currently do not support), we will get here for wide-decimals.

File: dbcon/joblist/lbidlist.cpp
Comment (tntnatbry) [reply dbcon/joblist/lbidlist.cpp:972]: Linker complains if I remove this template instantiation. Reason for this is the client of GetMinMax() is tuple-bps.cpp, which is a separate translation unit.

File: dbcon/joblist/rowestimator.cpp
Comment (tntnatbry) [reply dbcon/joblist/rowestimator.cpp:149]: This is intentional. The code at the end of this function is this:

```
    else
    {
        ret = max - min + 1;
    }

    if (ret > (T) fRowsPerExtent)
    {
        ret = fRowsPerExtent;
    }

    return ret;
```

So the type of ret has to be same as max/min, since we perform arithmetic on these values. Subsequently, `if (ret > (T) fRowsPerExtent)` ensures the maximum value returned fits in an uint32_t.

File: dbcon/joblist/lbidlist.cpp
Comment (drrtuy) [reply dbcon/joblist/lbidlist.cpp:711]: And I c that we don't treat value based on the dt used. Should't we have the same cimparison as in preceding blocks?

File: dbcon/joblist/rowestimator.cpp
Comment (tntnatbry) [reply dbcon/joblist/rowestimator.cpp]: I agree, this cast is not required. I am updating this.

File: dbcon/joblist/lbidlist.cpp
Comment (tntnatbry) [reply dbcon/joblist/lbidlist.cpp]: I agree. A direct cast to int128_t* should work just fine. I am changing this.

File: dbcon/joblist/rowestimator.cpp
Comment (tntnatbry) [reply dbcon/joblist/rowestimator.cpp]: Good catch. I missed the cast to __float128 here and in the next case, fixing this.

File: dbcon/joblist/tuple-bps.cpp
Comment (tntnatbry) [reply dbcon/joblist/tuple-bps.cpp]: Agreed. Changing this.

File: dbcon/joblist/tuple-bps.cpp
Comment (tntnatbry) [reply dbcon/joblist/tuple-bps.cpp:1706]: That's a million dollar question you have asked, I wish I had the answer. But I take part of the blame :)

File: dbcon/mysql/ha_mcs_partition.cpp
Comment (tntnatbry) [reply dbcon/mysql/ha_mcs_partition.cpp]: Done. I've replaced the hard coded value with utils::MAXLENGTH16BYTES

**Comment by tntnatbry** [comment]
Changes have been pushed based on the feedback, please review.

File: dbcon/joblist/lbidlist.cpp
Comment (tntnatbry) [reply dbcon/joblist/lbidlist.cpp:711]: Not sure I understand the last comment. The preceding else block is for unsigned types, for which the comparison involves first casting to the unsigned equivalent of either in128_t or int64_t. For the last else condition (signed datatypes), we compare the values directly.

**Review by drrtuy — APPROVED** [review]

**Pull request merged**
